### ISSUE TITLE
Improve performance of CA2000 (DisposeObjectsBeforeLosingScope)

### ISSUE THREAD

**Issue by sharwell** [dotnet/roslyn-analyzers#4915]
### Analyzer

**Diagnostic ID**: [CA2000](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2000): `Dispose objects before losing scope`

### Analyzer source

**Version**: 5.0.4-preview1.21126.5

### Describe the bug

CA2000 has been disabled in dotnet/roslyn-analyzers due to its pronounced impact on build times. The underlying performance issue(s) should be investigated and resolved so it can be enabled again.

### Steps To Reproduce

1. Revert d07ec24bc4dcfdbb37b390264c87c83fbb87ac77 to enable CA2000
1. Rebuild **RoslynAnalyzers.sln** with `/bl /p:ReportAnalyzer=true`
1. Open the **Build.binlog** to review the analyzer summary

### Expected behavior

CA2000 should not be responsible for significant analyzer time during the build.

### Actual behavior

CA2000 is responsible for the majority of all analyzer time in the build.

### Additional context

CA2000 was disabled in #4862.

**Comment by Youssef1313** [comment]
This analyzer had performance performance improvements in the past.
See https://github.com/dotnet/roslyn-analyzers/issues/3738

**Comment by sharwell** [comment]
The observations that led to this issue were based on a commit from just a few days ago.

**Comment by Youssef1313** [comment]
Yep I didn't mean this was fixed. But noting a *similar* issue just for reference.

**Comment by DS-AdamMilazzo** [comment]
The impact of this on our build is pretty huge:
```
  644.522   56      Microsoft.NetCore.Analyzers.Runtime.DisposeObjectsBeforeLosingScope (CA2000)
  3.613   <1      Microsoft.CodeQuality.Analyzers.QualityGuidelines.ValidateArgumentsOfPublicMethods (CA1062)
  0.122   <1      Microsoft.NetCore.Analyzers.Security.DoNotHardCodeCertificate (CA5403)
 ...
```
That's almost 11 minutes waiting for this one analyzer. This and CA1508 (see https://github.com/dotnet/roslyn-analyzers/issues/7125) are responsible for 99.8% of the build time.

**Comment by ImrePyhvel** [comment]
I concur that it is too costly to be of practical use. In our project:

> time(s)       %       Analyzer
> 1563.166   99      Microsoft.NetCore.Analyzers.Runtime.DisposeObjectsBeforeLosingScope (CA2000)

All other analyzers combined is <1%. We turned it off to get 6x faster builds.

I do understand that deep correctness comes at a cost and optimizing has increasin costs and diminishing returns... but perhaps having some additional "simplified" analyzer with reasonable perf to cover 95% of cases could be the middle ground? Give devs the choice depending on project & build.
